cpp/src/arrow/compute/registry.cc

```diff
@@ -34,78 +34,144 @@ namespace compute {
 
 class FunctionRegistry::FunctionRegistryImpl {
  public:
-  Status AddFunction(std::shared_ptr<Function> function, bool allow_overwrite) {
-#ifndef NDEBUG
-    // This validates docstrings extensively, so don't waste time on it
-    // in release builds.
-    RETURN_NOT_OK(function->Validate());
-#endif
-
-    std::lock_guard<std::mutex> mutation_guard(lock_);
+  explicit FunctionRegistryImpl(FunctionRegistryImpl* parent = NULLPTR)
+      : parent_(parent) {}
+  virtual ~FunctionRegistryImpl() {}
+
+  virtual Status CanAddFunction(std::shared_ptr<Function> function,
+                                bool allow_overwrite) {
+    return (parent_ == nullptr ? Status::OK()
+                               : parent_->CanAddFunction(function, allow_overwrite)) &
+           DoAddFunction(function, allow_overwrite, /*add=*/false);
```

Comment:
```suggestion
    if (parent_ != nullptr) {
        RETURN_NOT_OK(parent_->CanAddFunction(function, allow_overwrite)));
    }
    return DoAddFunction(function, allow_overwrite, /*add=*/false);
```

Now that it seems we won't be able to use `&&` we should probably move away from this more compact pattern and more explicitly short-circuit.